// @genkit-ai/dotprompt/lib/template.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var template_exports = {};
__export(template_exports, {
  compile: () => compile
});
module.exports = __toCommonJS(template_exports);
var import_handlebars = __toESM(require("handlebars"));
const Promptbars = import_handlebars.default.create();
function jsonHelper(serializable, options) {
  return new Promptbars.SafeString(
    JSON.stringify(serializable, null, options.hash.indent || 0)
  );
}
Promptbars.registerHelper("json", jsonHelper);
function roleHelper(role) {
  return new Promptbars.SafeString(`<<<dotprompt:role:${role}>>>`);
}
Promptbars.registerHelper("role", roleHelper);
function mediaHelper(options) {
  return new Promptbars.SafeString(
    `<<<dotprompt:media:url ${options.hash.url}${options.hash.contentType ? ` ${options.hash.contentType}` : ""}>>>`
  );
}
Promptbars.registerHelper("media", mediaHelper);
const ROLE_REGEX = /(<<<dotprompt:role:[a-z]+)>>>/g;
function toMessages(renderedString) {
  let currentMessage = {
    role: "user",
    source: ""
  };
  const messageSources = [currentMessage];
  for (const piece of renderedString.split(ROLE_REGEX).filter((s) => s.trim() !== "")) {
    if (piece.startsWith("<<<dotprompt:role:")) {
      const role = piece.substring(18);
      if (currentMessage.source) {
        currentMessage = { role, source: "" };
        messageSources.push(currentMessage);
      } else {
        currentMessage.role = role;
      }
    } else {
      currentMessage.source += piece;
    }
  }
  return messageSources.map((m) => ({
    role: m.role,
    content: toParts(m.source)
  }));
}
const PART_REGEX = /(<<<dotprompt:(?:media:url|section).*?)>>>/g;
function toParts(source) {
  const parts = [];
  const pieces = source.split(PART_REGEX).filter((s) => s.trim() !== "");
  for (let i = 0; i < pieces.length; i++) {
    const piece = pieces[i];
    if (piece.startsWith("<<<dotprompt:media:")) {
      const [_, url, contentType] = piece.split(" ");
      const part = { media: { url } };
      if (contentType)
        part.media.contentType = contentType;
      parts.push(part);
    } else if (piece.startsWith("<<<dotprompt:section")) {
      const [_, sectionType] = piece.split(" ");
      i++;
      const text = pieces[i];
      const part = { text, metadata: { purpose: sectionType } };
      parts.push(part);
    } else {
      parts.push({ text: piece });
    }
  }
  return parts;
}
function compile(source, metadata) {
  const renderString = Promptbars.compile(source, {
    knownHelpers: {
      json: true,
      section: true,
      media: true,
      role: true,
      history: true
    },
    knownHelpersOnly: true
  });
  return (input, options) => {
    const renderedString = renderString(input, {
      data: {
        metadata: { prompt: metadata, context: (options == null ? void 0 : options.context) || null }
      }
    });
    return toMessages(renderedString);
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  compile
});
