// @genkit-ai/vertexai/lib/predict.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var predict_exports = {};
__export(predict_exports, {
  predictModel: () => predictModel
});
module.exports = __toCommonJS(predict_exports);
var import_core = require("@genkit-ai/core");
function endpoint(options) {
  return `https://${options.location}-aiplatform.googleapis.com/v1/projects/${options.projectId}/locations/${options.location}/publishers/google/models/${options.model}:predict`;
}
const ACCESS_TOKEN_TTL = 50 * 60 * 1e3;
function predictModel(auth, { location, projectId }, model) {
  let accessToken;
  let accessTokenFetchTime = 0;
  return (instances, parameters) => __async(this, null, function* () {
    const fetch = (yield import("node-fetch")).default;
    if (!accessToken || accessTokenFetchTime + ACCESS_TOKEN_TTL < Date.now()) {
      accessToken = yield auth.getAccessToken();
      accessTokenFetchTime = Date.now();
    }
    const req = {
      instances,
      parameters: parameters || {}
    };
    const response = yield fetch(
      endpoint({
        projectId,
        location,
        model
      }),
      {
        method: "POST",
        body: JSON.stringify(req),
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
          "X-Goog-Api-Client": import_core.GENKIT_CLIENT_HEADER
        }
      }
    );
    if (!response.ok) {
      throw new Error(
        `Error from Vertex AI predict: HTTP ${response.status}: ${yield response.text()}`
      );
    }
    return yield response.json();
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  predictModel
});
