// @genkit-ai/firebase/lib/firestoreTraceStore.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-DJRN6NKF.mjs";
import { logger } from "@genkit-ai/core/logging";
import {
  SpanDataSchema,
  TraceDataSchema
} from "@genkit-ai/core/tracing";
import { Firestore } from "@google-cloud/firestore";
import { randomUUID } from "crypto";
const DOC_MAX_SIZE = 1e6;
class FirestoreTraceStore {
  constructor(params = {}) {
    this.collection = params.collection || "genkit-traces";
    this.databaseId = params.databaseId || "(default)";
    this.expireAfterDays = params.expireAfterDays || 14;
    this.db = new Firestore({
      databaseId: this.databaseId,
      ignoreUndefinedProperties: true
    });
  }
  save(traceId, traceData) {
    return __async(this, null, function* () {
      const expireAt = Date.now() + this.expireAfterDays * 24 * 60 * 60 * 1e3;
      const traceInfo = __spreadProps(__spreadValues({}, traceData), {
        expireAt,
        spans: {}
      });
      const start = Date.now();
      const batches = rebatchSpans(traceData);
      let batchWrite = this.db.batch();
      batchWrite.set(
        this.db.collection(this.collection).doc(traceId),
        traceInfo,
        { merge: true }
      );
      batches.forEach((batch) => {
        batchWrite.create(
          this.db.collection(this.collection).doc(traceId).collection("spans").doc(randomUUID()),
          {
            expireAt,
            spans: batch
          }
        );
      });
      yield batchWrite.commit();
      logger.debug(
        `saved trace ${traceId}, ${Object.keys(traceData.spans).length} span(s) (${Date.now() - start}ms)`
      );
    });
  }
  load(traceId) {
    return __async(this, null, function* () {
      const [traceInfo, spanBatches] = yield Promise.all([
        this.db.collection(this.collection).doc(traceId).get(),
        this.db.collection(this.collection).doc(traceId).collection("spans").get()
      ]);
      const spans = {};
      spanBatches.forEach((batch) => {
        const spansBatch = batch.data().spans;
        Object.keys(spansBatch).forEach((key) => {
          spans[key] = SpanDataSchema.parse(spansBatch[key]);
        });
      });
      return TraceDataSchema.parse(__spreadProps(__spreadValues({}, traceInfo.data()), {
        spans
      }));
    });
  }
  list(query) {
    return __async(this, null, function* () {
      const limit = (query == null ? void 0 : query.limit) || 10;
      let fsQuery = this.db.collection(this.collection).orderBy("startTime", "desc");
      if (query == null ? void 0 : query.continuationToken) {
        fsQuery = fsQuery.startAfter(parseInt(query.continuationToken));
      }
      fsQuery = fsQuery.limit(limit);
      const data = yield fsQuery.get();
      const lastVisible = data.docs[data.docs.length - 1];
      return {
        traces: data.docs.map((d) => d.data()),
        continuationToken: data.docs.length === limit ? `${lastVisible.data().startTime}` : void 0
      };
    });
  }
}
function rebatchSpans(traceData) {
  const batches = [];
  let lastBatchRunningSize = 0;
  for (const span of Object.values(traceData.spans)) {
    let estimatedSpanSize = estimateSpanSize(span);
    if (estimatedSpanSize >= DOC_MAX_SIZE) {
      logger.warn(
        `Truncating data for trace ${traceData.traceId} span ${span.spanId}`
      );
      truncateSpanData(span);
      estimatedSpanSize = estimateSpanSize(span);
    }
    if (lastBatchRunningSize + estimatedSpanSize < DOC_MAX_SIZE) {
      if (batches.length === 0) {
        batches.push({});
      }
    } else {
      batches.push({});
      lastBatchRunningSize = 0;
    }
    lastBatchRunningSize += estimatedSpanSize;
    batches[batches.length - 1][span.spanId] = span;
  }
  return batches;
}
function estimateSpanSize(span) {
  if (Object.values(span.attributes).length === 0) {
    return 0;
  }
  return Object.values(span.attributes).map((attr) => attr.toString().length).reduce((a, b) => a + b);
}
function truncateSpanData(span) {
  span.truncated = true;
  delete span.attributes["genkit:output"];
  if (estimateSpanSize(span) < DOC_MAX_SIZE) {
    return;
  }
  delete span.attributes["genkit:input"];
  if (estimateSpanSize(span) < DOC_MAX_SIZE) {
    return;
  }
  span.attributes = {};
  return;
}
export {
  FirestoreTraceStore,
  rebatchSpans
};
