// @genkit-ai/ai/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  Candidate: () => import_generate.Candidate,
  GenerateResponse: () => import_generate.GenerateResponse,
  Message: () => import_generate.Message,
  asTool: () => import_tool.asTool,
  definePrompt: () => import_prompt.definePrompt,
  defineTool: () => import_tool.defineTool,
  evaluate: () => import_evaluator.evaluate,
  evaluatorRef: () => import_evaluator.evaluatorRef,
  generate: () => import_generate.generate,
  generateStream: () => import_generate.generateStream,
  index: () => import_retriever.index,
  indexerRef: () => import_retriever.indexerRef,
  renderPrompt: () => import_prompt.renderPrompt,
  retrieve: () => import_retriever.retrieve,
  retrieverRef: () => import_retriever.retrieverRef,
  toGenerateRequest: () => import_generate.toGenerateRequest
});
module.exports = __toCommonJS(src_exports);
var import_evaluator = require("./evaluator.js");
var import_generate = require("./generate.js");
var import_prompt = require("./prompt.js");
var import_retriever = require("./retriever.js");
var import_tool = require("./tool.js");
__reExport(src_exports, require("./types.js"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Candidate,
  GenerateResponse,
  Message,
  asTool,
  definePrompt,
  defineTool,
  evaluate,
  evaluatorRef,
  generate,
  generateStream,
  index,
  indexerRef,
  renderPrompt,
  retrieve,
  retrieverRef,
  toGenerateRequest,
  ...require("./types.js")
});
