// @genkit-ai/flow/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  FirestoreStateStore: () => import_firestoreStateStore.FirestoreStateStore,
  Flow: () => import_flow.Flow,
  FlowInvokeEnvelopeMessageSchema: () => import_types.FlowInvokeEnvelopeMessageSchema,
  FlowStateExecutionSchema: () => import_core.FlowStateExecutionSchema,
  OperationSchema: () => import_core.OperationSchema,
  defineFlow: () => import_flow.defineFlow,
  getFlowAuth: () => import_utils.getFlowAuth,
  run: () => import_steps.run,
  runAction: () => import_steps.runAction,
  runFlow: () => import_flow.runFlow,
  runMap: () => import_steps.runMap,
  startFlowsServer: () => import_flow.startFlowsServer,
  streamFlow: () => import_flow.streamFlow
});
module.exports = __toCommonJS(src_exports);
var import_core = require("@genkit-ai/core");
var import_firestoreStateStore = require("./firestoreStateStore.js");
var import_flow = require("./flow.js");
var import_steps = require("./steps.js");
var import_types = require("./types.js");
var import_utils = require("./utils.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FirestoreStateStore,
  Flow,
  FlowInvokeEnvelopeMessageSchema,
  FlowStateExecutionSchema,
  OperationSchema,
  defineFlow,
  getFlowAuth,
  run,
  runAction,
  runFlow,
  runMap,
  startFlowsServer,
  streamFlow
});
